src/llfs/track_fds.cpp

```diff
@@ -0,0 +1,188 @@
+//#=##=##=#==#=#==#===#+==#+==========+==+=+=+=+=+=++=+++=+++++=-++++=-+++++++++++
+//
+// Part of the LLFS Project, under Apache License v2.0.
+// See https://www.apache.org/licenses/LICENSE-2.0 for license information.
+// SPDX short identifier: Apache-2.0
+//
+//+++++++++++-+-+--+----- --- -- -  -  -   -
+
+#include <llfs/track_fds.hpp>
+//
+
+#include <llfs/logging.hpp>
+
+#include <batteries/env.hpp>
+#include <batteries/suppress.hpp>
+
+#include <filesystem>
+#include <memory>
+#include <mutex>
+#include <unordered_map>
+
+#ifdef __linux__
+//
+#include <dirent.h>
+#include <sys/types.h>
+//
+#endif
+
+namespace llfs {
+
+namespace {
+
+const char* kVarName = "LLFS_TRACK_FDS";
+
+struct TrackFdState {
+  std::atomic<bool> enabled{batt::getenv_as<int>(kVarName).value_or(0) != 0};
+  std::mutex mutex;
+  std::unordered_map<int, std::unique_ptr<boost::stacktrace::stacktrace>> traces;
+
+  //+++++++++++-+-+--+----- --- -- -  -  -   -
+
+  static TrackFdState& instance() noexcept
+  {
+    // Intentionally leaked.
+    //
+    static TrackFdState* const instance_ = new TrackFdState{};
+
+    return *instance_;
+  }
+};
+
+}  //namespace
+
+//==#==========+==+=+=++=+++++++++++-+-+--+----- --- -- -  -  -   -
+//
+std::set<int> get_open_fds()
+{
+  std::set<int> result;
+  const char* dirpath = "/proc/self/fd";
+
+  //+++++++++++-+-+--+----- --- -- -  -  -   -
+#ifdef __linux__
+  DIR* p_dir = opendir(dirpath);
+
+  if (!p_dir) {
+    LLFS_LOG_ERROR() << "Could not open " << dirpath;
+  } else {
+    auto on_scope_exit = batt::finally([&] {
+      closedir(p_dir);
+    });
+
+    const int p_dir_fd = dirfd(p_dir);
+
+    isize name_max = pathconf(dirpath, _PC_NAME_MAX);
+    if (name_max == -1) { /* Limit not defined, or error */
+      name_max = 255;     /* Take a guess */
+    }
+    const usize len = offsetof(struct dirent, d_name) + name_max + 1;
+    void* const entry_buffer = malloc(len);
+    auto on_scope_exit2 = batt::finally([&] {
+      free(entry_buffer);
+    });
+
+    for (;;) {
+      struct dirent* p_entry = nullptr;
+
+      BATT_SUPPRESS_IF_GCC("-Wdeprecated-declarations")
+      const int retval = readdir_r(p_dir, (struct dirent*)entry_buffer, &p_entry);
+      BATT_UNSUPPRESS_IF_GCC()
+
+      // Check for errors.
+      //
+      if (retval != 0) {
+        LLFS_LOG_ERROR() << "readdir_r returned: " << retval;
+        break;
+      }
+
+      // If p_entry is null, we are done!
+      //
+      if (!p_entry) {
+        break;
+      }
+
+      // If the entry name parses as an integer *and* it is not the open directory we are scanning,
+      // then add it to the result set.
+      //
+      std::optional<int> fd = batt::from_string<int>(p_entry->d_name);
+      if (fd && *fd != p_dir_fd) {
+        result.emplace(*fd);
+      }
+    }
+  }
+
+  //+++++++++++-+-+--+----- --- -- -  -  -   -
+#else
```

Comment:
Add `TODO: Make Windows compatible`. May as well start tracking where we'll need to go back and make changes for Windows. Could use a unique identifier so we can easily search for these locations in the future.

Reply:
Good idea!  Done.